x/ibc/callbacks/keeper/keeper.go

```diff
@@ -0,0 +1,429 @@
+// 5. Calls the contract's onPacketAcknowledgement function with packet details
+// 6. Manages gas consumption and validates gas limits
+//
+// Returns:
+//   - error: Returns nil on success, or an error if any step fails including:
+//   - Packet data unmarshaling errors
+//   - Invalid callback data or unexpected calldata presence
+//   - Address parsing failures
+//   - Contract validation failures (non-existent or no code)
+//   - ABI loading errors
+//   - EVM execution errors
+//   - Gas limit exceeded errors
+//
+// Contract Requirements:
+//   - Must implement onPacketAcknowledgement(string calldata sourceChannel, string calldata sourcePort,
+//     uint64 sequence, bytes calldata data, bytes calldata acknowledgement) function
+//   - Should handle both successful and failed acknowledgements appropriately
+func (k ContractKeeper) IBCOnAcknowledgementPacketCallback(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+	contractAddress,
+	packetSenderAddress string,
+	version string,
+) error {
+	data, err := transfertypes.UnmarshalPacketData(packet.GetData(), version, "")
+	if err != nil {
+		return err
+	}
+
+	cbData, isCbPacket, err := callbacktypes.GetCallbackData(data, version, packet.GetDestPort(), ctx.GasMeter().GasRemaining(), ctx.GasMeter().GasRemaining(), callbacktypes.SourceCallbackKey)
+	if err != nil {
+		return err
+	}
+	if !isCbPacket {
+		return nil
+	}
+
+	// `ProcessCallback` in IBC-Go overrides the infinite gas meter with a basic gas meter,
+	// so we need to generate a new infinite gas meter to run the EVM executions on.
+	// Skipping this causes the EVM gas estimation function to deplete all Cosmos gas.
+	// We re-add the actual EVM call gas used to the original context after the call is complete
+	// with the gas retrieved from the EVM message result.
+	cachedCtx, writeFn := ctx.CacheContext()
+	cachedCtx = evmante.BuildEvmExecutionCtx(cachedCtx).
+		WithGasMeter(types2.NewInfiniteGasMeterWithLimit(cbData.CommitGasLimit))
+
+	if len(cbData.Calldata) != 0 {
+		return errorsmod.Wrap(types.ErrInvalidCalldata, "acknowledgement callback data should not contain calldata")
+	}
+
+	sender, err := utils.HexAddressFromBech32String(packetSenderAddress)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to parse packet sender address %s", packetSenderAddress)
+	}
+
+	contractAddr := common.HexToAddress(contractAddress)
+	contractAccount := k.evmKeeper.GetAccountOrEmpty(ctx, contractAddr)
+
+	// Check if the contract address contains code.
+	// This check is required because if there is no code, the call will still pass on the EVM side,
+	// but it will ignore the calldata and funds may get stuck.
+	if !contractAccount.IsContract() {
+		return errorsmod.Wrapf(types.ErrCallbackFailed, "provided contract address is not a contract: %s", contractAddr)
+	}
+
+	abi, err := callbacksabi.LoadABI()
+	if err != nil {
+		return err
+	}
```

Comment:
This is mostly the same across multiple callbacks right? Can we just refactor this into a function and call it in each one?